### ISSUE TITLE
Parts with embedded Textures

### ISSUE THREAD

**Issue by tobblx** [ScanMountGoat/ldr_tools_blender#31]
Hello,
I have another big feature-request for you on my wishlist :-)
For creating the ldraw Models I use Lego "Studio 2.0" It contains lots of unofficial parts in the directory "...\Lego_Studio 2.0\ldraw\UnOfficial\parts"
Many of them have Base64-encoded embedded PNG textures. Those textures are projected on the parts. (I dont't find the source where this is specified)

Example:

[https://www.bricklink.com/v2/catalog/catalogitem.page?P=25269pb030#T=C&C=2](https://www.bricklink.com/v2/catalog/catalogitem.page?P=25269pb030#T=C&C=2)

_0 Tile, Round 1 x 1 Quarter with Red Watermelon Pattern
0 Name: 25269pb002.dat
0 Author: 
0 BFC CERTIFY CCW
0 PE_TEX_PATH 0 1
**0 PE_TEX_INFO 0.5000 0.9500 0.5000 0.0000 0.0000 -1.6100 0.0000 2.2000 0.0000 1.9100 0.0000 0.0000 -10.0000 10.0000 10.0000 -10.0000 iVBORw0KGgoAAAANSUhE . . . CHDKdcQAAAAASUVORK5CYII=**
1 16 0 0 0 1 0 0 0 1 0 0 0 1 25269.dat_

You can search those parts in the parts-folder by the content "PE_TEX_INFO"

It would be great, if those textures could be parsed/extracted from the parts when importing the ldraw files.
When creating the materials, an additional texture-node would be needed, using the cached texture from the parts-File
UVs for those parts would also be needed to be calculated. Maybe it could be done in the shader directly somehow without UVs, but correct UVs on the Meshes would be much better to handle. And for using the models in game-engines later eventually, without UVs it would be much worse.

Example:
![Pizza1](https://github.com/ScanMountGoat/ldr_tools_blender/assets/121121299/41ab7064-68f0-4688-b1f6-a0e8600df86d)
![grafik](https://github.com/ScanMountGoat/ldr_tools_blender/assets/121121299/9ff9d948-b00a-4806-8bfe-542a25db3cb1)
![grafik](https://github.com/ScanMountGoat/ldr_tools_blender/assets/121121299/ef5cd53a-3df1-4319-a1a2-eb74881aec0e)

For my Blender-project, i created a script, which just makes a planar UV-projection for the pizza and melon textures on the needed parts. (see script below) But the real info how to map the UVs is somewhere hidden in the "PE_TEX_INFO"

I think it's a huge work to put all those things together for the importer, but lots of those parts could then be used without extra scripts and without extracting the textures manually from the part-Files.

```
import bpy

def set_uv_from_vertex_coordinates():
    # Get the active object
    obj = bpy.context.active_object

    # Ensure the object is a mesh
    if obj is None or obj.type != 'MESH':
        print("Please select a mesh object.")
        return

    # Switch to Object mode to apply changes to UVs
    bpy.ops.object.mode_set(mode='OBJECT')

    # Ensure the object has UV map
    if not obj.data.uv_layers:
        obj.data.uv_layers.new(name="UVMap")

    # Get the UV map
    uv_layer = obj.data.uv_layers.active.data

    # Access the mesh data
    mesh = obj.data
    uv_map = mesh.uv_layers.active.data

    # Find the min and max X, Y coordinates
    min_x = min(vert.co.x for vert in mesh.vertices)
    max_x = max(vert.co.x for vert in mesh.vertices)
    min_z = min(vert.co.z for vert in mesh.vertices)
    max_z = max(vert.co.z for vert in mesh.vertices)

    range_x = max_x - min_x
    range_z = max_z - min_z

    # Create a vertex to UV map
    vertex_to_uv_map = {}
    for poly in mesh.polygons:
        for loop_index in poly.loop_indices:
            loop_vert_index = mesh.loops[loop_index].vertex_index
            vertex_to_uv_map[loop_index] = loop_vert_index

    # Set UV coordinates based on vertex coordinates
    for loop_index, vert_index in vertex_to_uv_map.items():
        vertex = mesh.vertices[vert_index]
        u = (vertex.co.x - min_x) / range_x
        v = (vertex.co.z - min_z) / range_z
        uv_map[loop_index].uv = (0, 0)
        if "25269pb002" in obj.name:
            uv_map[loop_index].uv = (u, v)
        if "bl_25269pb030" in obj.name:
            uv_map[loop_index].uv = (v, -u)
        if "25269pb003" in obj.name:
            uv_map[loop_index].uv = (-v, u)

    print("UV coordinates set from vertex coordinates. für "+obj.name)

# Run the function
set_uv_from_vertex_coordinates()


```

**Comment by cuddlyogre** [comment]
The way Stud.io does its texmaps is not terribly straightforward. I've figured out most of it, so of you're looking for the logic, you can reference https://github.com/cuddlyogre/ExportLDraw/blob/master/pe_texmap.py

The only thing missing is handling cases where the texmap is mapped via box projection (I think), but it is very uncommon to run across parts with textures applied in a way that requires it. In fact, the only way I could get it to generate those kinds of texmaps was to intentionally apply textures haphazardly, so I imagine it would be an uncommon case.

**Comment by ScanMountGoat** [comment]
What are the values in the PE_TEX_INFO line supposed to represent other than the base64 encoded image data?

**Comment by cuddlyogre** [comment]
They are PNGs encoded in base64. I don't believe any other formats are supported in PartDesigner. Unline the LDraw spec, the base64 string is not split int multiple lines. The whole things is at the end of the line in full, which depending on the image can make working in the file difficult.

**Comment by ScanMountGoat** [comment]
I'm referring to the part before that like `0.5000 0.9500 0.5000 0.0000 0.0000 -1.6100 0.0000 2.2000 0.0000 1.9100 0.0000 0.0000 -10.0000 10.0000 10.0000 -10.0000`.

**Comment by cuddlyogre** [comment]
I believe that is the transformation of the box for the box projection, which I haven't figured out.

**Comment by ScanMountGoat** [comment]
You can see the exact values used by Studio using a GPU debugger like RenderDoc. Matching the UVs hopefully won't be super complicated. 
![image](https://github.com/ScanMountGoat/ldr_tools_blender/assets/23301691/c6dcf9de-1a63-4649-a921-3c0ce4a396c5)

**Comment by cuddlyogre** [comment]
Oh dang. I didn't know about that tool. Thank you for that.

**Comment by ScanMountGoat** [comment]
It looks like the PE_TEX_INFO commands can appear with or without the 16 float parameter values. It's not super obvious what the parameters are supposed to do. There's also the PE_TEX_PATH commands to figure out.

I'm prioritizing other projects at the moment, but I would be happy to review code or provide guidance for someone else that wants to look into this. One approach to get started is to export different parts from PartDesigner as ldraw and inspect the TEXCOORD0 values in RenderDoc as a reference.

**Comment by The0x539** [comment]
I believe I have this figured out in cuddlyogre's plugin. It's mixed in with a bunch of other changes, some incomplete, but see [this branch](https://github.com/The0x539/ExportLDraw/tree/studio-colors) for reference. I didn't know about *this* project before today - I'll try my hand at porting what I've done over.

**Comment by The0x539** [comment]
Update: For parts that include UV data in their polygons, this will require modifying `weldr` to be able to represent the extra UV vertices that go on `3` and `4` commands. "Projected" textures shouldn't require any such changes, although `ldr_tools` is definitely not the ideal place to be parsing any of the proprietary commands.

**Comment by ScanMountGoat** [comment]
> For parts that include UV data in their polygons, this will require modifying weldr to not discard the extra UV vertices that go on 3 and 4 command

Do you have an example file or what the extended command would look like? I can update my fork with any needed changes to parsing.

> I'll try my hand at porting what I've done over

That would be great. Is there a specific part of the code you can link to that handles the tex maps? I should be able to give some guidance on where the parts of the implementation should go. I would expect most of the changes to go in ldr_tools itself.

**Comment by The0x539** [comment]
> Do you have an example file or what the extended command would look like?
If your parts library has it, look at `3297tex.dat`, or more specifically, `slopetex.dat`.
The data is just six (or eight) extra coordinates that should be parsed as a new `uvs: Option<[Vec2; 3]>` (or `[Vec2; 4]`) field on `TriangleCmd` and `QuadCmd`.

> Is there a specific part of the code you can link to that handles the tex maps?
`pe_texmap.py` and some parts of `ldraw_meta.py`. That said, I've already gotten started. I find Rust codebases *much* easier to navigate than anything else due to the power of local reasoning (the FFI barrier also helps identify what functions *won't* do).
The task still feels a bit daunting as a whole, and it's tricky to figure out where/how to break it down into smaller problems, but I think I understand most of what needs to be done.

See [my `textures` branch](https://github.com/The0x539/ldr_tools_blender/tree/textures) for what I've done so far.

**Comment by ScanMountGoat** [comment]
> The data is just six (or eight) extra coordinates that should be parsed as a new uvs: Option<[Vec2; 3]> (or [Vec2; 4]) field on TriangleCmd and QuadCmd.

That's what I was thinking as well. It looks like you can just do the parsing manually for now since the format is relatively simple. I'll update my fork of weldr later on with any needed changes.

It looks like you have a good start on your branch so far. Let me know if you run into any issues.

**Comment by The0x539** [comment]
I can manually parse most of the data, but as far as I can tell, I need weldr's help for the tri/quad UVs, because you only get the original line for "comments" by the time I use it in `append_geometry`. When parsing the tri/quad commands, the UV text is discarded with no way to recover it.

**Comment by ScanMountGoat** [comment]
It should be working now if you use `weldr = { git = "https://github.com/ScanMountGoat/weldr", rev = "a26a0c7" }` in your `Cargo.toml`. This adds support for the optional UVs on triangle and quad commands.

**Comment by The0x539** [comment]
Yep, that did the trick - thanks!. I'm passing some meaningful data up to Python and loading the images. Setting the UVs might be trickier than I expected, because I don't yet understand how they're represented in the API, and ExportLDraw does it using BMesh. Still, some good progress today.

**Comment by ScanMountGoat** [comment]
Nice! Feel free to make a PR when you feel like the implementation is mostly working.

Blender's API docs are a little hard to follow. This should work with some adjustment when you get to that part. I would use a numpy array for the UVs similar to how other attributes are handled in ldr_tools_py. This is much faster than lists and also makes it easy to reindex the UVs.

```python
uv_layer = blender_mesh.uv_layers.new(name=name)
loop_uvs = vertex_uvs[vertex_indices]
uv_layer.data.foreach_set("uv", loop_uvs.reshape(-1))
```

**Comment by The0x539** [comment]
I have achieved successful texture import with my placeholder UVs. A complete implementation should not be too far off.

**Comment by The0x539** [comment]
I think I'm stuck.
I have a test file that looks like this in Studio and when imported with my fork of cuddlyogre's plugin:
![image](https://github.com/user-attachments/assets/5494b77a-9d79-40c6-b5b0-fe6ad4a3cced)

My `textures` branch for this plugin *almost* gets it right: 
![image](https://github.com/user-attachments/assets/f99bab74-63f2-4f0c-ad0e-8b67c095d084)

I can't figure out what's going wrong with the 3x3 slope. (Ignore the studs, that's just because I haven't done bounding-box filtering yet, because it's historically made troubleshooting harder on the faces that *are* supposed to be textured.)

It's probably some combination of axis mirroring and winding-order handling, but I cannot identify the correct combination of things to do. When I try to print any of the data I'm working with, the signal-to-noise ratio is such that I don't know what output relates to the actual face of interest.

[pe_texmap test case.zip](https://github.com/user-attachments/files/16986944/pe_texmap.test.case.zip)

**Comment by ScanMountGoat** [comment]
Is there a reason why you handle winding order inside `project_texture`? Applying the winding to the vertex indices should be equivalent to applying the winding to all attributes like positions and uvs. 

I would also try without welding vertices. The UVs also need to be indexed together with positions or stored separately for each vertex in each face (loop in blender). It looks like this is handled in your Python code, but it might be worth checking without welding.

**Comment by ScanMountGoat** [comment]
It fixes the issue on that specific part by negating both U and V. I'm not sure why this would be needed. I'll try and look into this later.
```rust
let mut uvs = vertices.map(|vert| {
    let u = (-vert.x - min.x) / diff.x;
    let v = (-vert.z - min.y) / diff.y;
    Vec2::new(u, v)
});
```

**Comment by The0x539** [comment]
> Is there a reason why you handle winding order inside `project_texture`?

Mostly because I don't really know what I'm doing. My ExportLDraw code for UV maps had some provisions for winding order separate from how the main code handled it, so I guessed that I'd need something analogous here.

> The UVs need to be indexed together with positions or stored separately for each vertex in each face (loop in blender). It looks like this is handled in your Python code, 

I would be surprised if this were a problem. Currently, the Rust code generates one "texmap" per face, which the Python code converts to one U coordinate and one V coordinate per loop. I was planning on replacing this with a more efficient format once I/we figure out all the math.

> I'll try and look into this later.

Thank you. Had I stumbled into something that seemed to work, I probably would have blindly moved forward with it, but I would *love* to be confident that the code is actually doing the correct thing.

> Are your texture transform matrices and point min/max the same in both addons?

Checking this was my first troubleshooting instinct, but I was unable to confirm one way or the other, because when I log the coordinate data, I have a very hard time telling which face/vertex/whatever is which. I barely managed to do that when I was comparing Studio's code (logging injected with a modding framework) with my first version, and in that case I had the reassurance that Studio's implementation was "correct" by definition.

**Comment by ScanMountGoat** [comment]
Do we know the high level operation being applied by the PE_TEX_INFO lines in those example files? There are examples for different projection techniques here: https://learn.foundry.com/modo/content/help/pages/shading_lighting/shader_items/projection_type_samples.html

The negation and element rearranging on the input matrix is a little suspicious. Is this changing the coordinate system? Ideally, this would be expressed using matrix multiplication or some other named operation to avoid any confusion.

> I have a very hard time telling which face/vertex/whatever is which

It looks like UVs are being saved to TEXCOORD1 in RenderDoc. The values are written to the vertex output as is, so this is likely calculated in C# somewhere. Changing the parameters for the PE_TEX_INFO line changes these values. Setting the last 4 to 0 results in inf, for example. It looks like -1, -1 is for vertices that shouldn't have UVs. RenderDoc will show the selected vertex if you click on lines in the mesh viewer.
![image](https://github.com/user-attachments/assets/de5322f2-198d-4c07-88e9-fd8a659f879e)

**Comment by ScanMountGoat** [comment]
I'm not sure if using the last 4 parameters as `(uv - min) / (max - min)` is correct. Using all 0's produces infinity as expected, but setting min and max to the same value does not. It also doesn't scale the UVs in a predictable way in RenderDoc. There are a lot of projection techniques that will produce similar results on flat surfaces, so it's possible the high level approach is wrong.

**Comment by The0x539** [comment]
I'll dig through the decompiled C# as soon as I get a chance - that should give us a few answers.

**Comment by The0x539** [comment]
From `LDrawTextureInfo.InitWithLine`:
```cs
Matrix4x4 matrix = default(Matrix4x4);
matrix.m03 = LDrawPrimitive.parseFloat(array[0], 1E-05f);
matrix.m13 = LDrawPrimitive.parseFloat(array[1], 1E-05f);
matrix.m23 = -LDrawPrimitive.parseFloat(array[2], 1E-05f);
matrix.m00 = float.Parse(array[3]);
matrix.m01 = float.Parse(array[4]);
matrix.m02 = -float.Parse(array[5]);
matrix.m10 = float.Parse(array[6]);
matrix.m11 = float.Parse(array[7]);
matrix.m12 = -float.Parse(array[8]);
matrix.m20 = -float.Parse(array[9]);
matrix.m21 = -float.Parse(array[10]);
matrix.m22 = float.Parse(array[11]);
matrix.m30 = 0f;
matrix.m31 = 0f;
matrix.m32 = 0f;
matrix.m33 = 1f;
this.m_pointMin = Vector2.zero;
this.m_pointMax = Vector2.zero;
this.m_pointMin.x = LDrawPrimitive.parseFloat(array[12], 1E-05f);
this.m_pointMin.y = LDrawPrimitive.parseFloat(array[13], 1E-05f);
this.m_pointMax.x = LDrawPrimitive.parseFloat(array[14], 1E-05f);
this.m_pointMax.y = LDrawPrimitive.parseFloat(array[15], 1E-05f);
this.m_pointDiff = this.m_pointMax - this.m_pointMin;
this.m_boxExtents = 0.5f * Vector3.one;
this.m_matrix = matrix;
this.m_matrixInverse = matrix.inverse;
this.m_textureImgData = Convert.FromBase64String(array[16]);
```
So that's where the matrix element negation originated from. I'm not sure what exactly you mean by rearranging, unless I misinterpreted the semantics of the `Mat4` constructor.

However, this is happening in Unity's coordinate space, which I *believe* is different from both LDraw and Blender, and I don't know whether (or how) that changes what exactly should happen here.


The `(uv - min) / (max - min)` comes from `LDrawTextureInfo.GetUV`:
```cs
Vector2 zero = Vector2.zero;
Vector2 vector = Vector3.zero;
vector.x = (point.x - this.m_pointMin.x) / this.m_pointDiff.x;
vector.y = (point.z - this.m_pointMin.y) / this.m_pointDiff.y;
zero.x = this.m_atlasMinUV.x + vector.x * this.m_atlasDiffUV.x;
zero.y = this.m_atlasMinUV.y + vector.y * this.m_atlasDiffUV.y;
return zero;
```
I'm not sure why `pointDiff` is cached, but I vaguely recall it might be something about not actually being equal to `max - min` at the site of usage. I can't find anything in the code to support this, though.

`project_texture` corresponds directly to `LDrawTextureInfo.InitMatrixWithTargetPartMatrix`. I'm glad the Rust version is so much simpler than dnSpy's decompiled C#, thanks to actual element-wise `abs` and `signum` operations.

Finally, if you have any interest in browsing the code yourself with dnSpy, this "call tree" with notes that I pieced together while I implemented my first version of texture projection may serve as a guide:
```
LDrawModel.initWithLines
	LDrawModel.CheckTexture
		LDrawTextureInfo.Create
			LDrawTextureInfo.InitWithLine

LDrawTextureAtlas.AssignUV
	LDrawTextureAtlas.GetTargetMeshInfoFromPath (leaf)
		descends a tree of LDrawTextureMeshInfo based on a PE_TEX_PATH
		keeps track of the cumulative matrix as it descends
		also keeps track of whether the mesh "is inverted"
	LDrawTextureMeshInfo.Flatten
		LDrawTextureMeshInfo.Flatten (recursion)
		converts a hierarchical structure of ldraw meshes into a flat mesh of vertices, edges, and faces
		the conversion is in-place via side effects
	LDrawTextureInfo.InitMatrixWithTargetPartMatrix (leaf)
		recomputes self.matrix and self.matrix_inverse
		first, matmul: target part matrix * this.matrix
		then, surgically extract the MAGNITUDE of the scale component and store it in self.box_extents
		self.matrix and self.matrix_inverse retain the SIGN of the scale component
	LDrawTextureInfo.AssignUVToMeshInfo
		LDrawTextureInfo.CollectVerticesInBoxExtents
			LDrawTextureInfo.IsUVOfFaceAlreadyAssigned (leaf)
				checks if any of the face's vertices have a positive U coordinate
			LDrawTextureInfo.CheckFaceNormalWithTextureNormal (leaf)
				computes a face's normal vector using the cross product of two of its edges
				takes the dot product with the texture normal
				returns true for substantially-positive dot product
			LDrawTextureInfo.IsIntersecting (leaf)
				seems to be some kind of triangle/cuboid intersection algorithm
		LDrawTextureInfo.CollectConnectVertices
			LDrawTextureInfo.IsUVOfFaceAlreadyAssigned (leaf)
			LDrawTextureInfo.CheckFaceNormalWithTextureNormal (leaf)
		LDrawTextureInfo.GetUV (leaf)
			does arithmetic with the input point and min/max/diff values
```

Even with all this information, I'm not sure what to do. This 3D transformation math is not my strong suit, and it seems like I mostly got lucky with getting my first implementation to work (or, well, work with the test cases I have handy).

**Comment by ScanMountGoat** [comment]
> I'm not sure what exactly you mean by rearranging

I found the code you mentioned in the decompiled assembly. The file doesn't store a row from the matrix since it's always (0 0 0 1). In games, I often see this as mat3x4 instead of putting the translation part first.

> However, this is happening in Unity's coordinate space, which I believe is different from both LDraw and Blender, and I don't know whether (or how) that changes what exactly should happen here.

This is important to double check. I don't see anything obviously wrong with your code after comparing with the decompiled output. Incorrect coordinate system handling could result in flipping or rotations by some multiple of 90 degrees. ldr_tools does all its math in LDraw's weird coordinate system. This is probably different than what Studio is doing.

I'm getting a more interesting failure case with the pizza slice mentioned at the top of this issue (bl_25269pb030.dat). It looks like the UVs are flipped. The face issue may be winding related. It also wouldn't surprise me if the UVs need to be flipped vertically in Blender with `1.0 - uv.y`, masking some other issue with the coordinate system handling.
![image](https://github.com/user-attachments/assets/f9886ea2-f90b-47c9-acbc-8ccc6cfc1e28)

**Comment by ScanMountGoat** [comment]
In `CheckFaceNormalWithTextureNormal`, Studio checks if the face points in the same general direction as the texture normal (0, -1, 0). This may give some clues about how they expect the coordinates to be set up for UV calculation.

**Comment by The0x539** [comment]
I made a slimmed-down version of `25269pb030.dat` with just the top faces (and a heavily `pngquant`'d texture) to hopefully make it easier for me to look at geometry data logs.
[25269test.dat](https://github.com/user-attachments/files/17017919/25269test.txt) (renamed to .txt because GitHub is being weird)

I observe that my implementation in `ExportLDraw` handles the pizza slice correctly.

> ldr_tools does all its math in LDraw's weird coordinate system. This is probably different than what Studio is doing.

This is now my primary suspect. `ExportLDraw` appears to have the Blender/LDraw transformation "ingrained" in its import mechanism: https://github.com/cuddlyogre/ExportLDraw/blob/9c32bc3e90fc9df4b7b94936d63e6d00b78c2e11/ldraw_node.py#L80C39-L80C65

Considering Studio and `ExportLDraw` both do such a thing, while your approach is "pure LDraw" until it gets to the part that actually talks to Blender, I'm going to guess that the correct approach involves _less_ "tampering" with the transformations.

**Comment by The0x539** [comment]
> In games, I often see this as mat3x4 instead of putting the translation part first.

It looks like `glam::f32::Affine3A` is the datatype for this here. I'll try to make use of it in the final version at minimum.

**Comment by ScanMountGoat** [comment]
I would just stick with `Mat4` for consistency. It would be helpful to add a comment for why a 4x4 matrix is initialized using only using 12 parameters.

**Comment by The0x539** [comment]
I removed the matrix element negation and tweaked some other stuff.

![image](https://github.com/user-attachments/assets/1c1f4f82-fc9f-4abb-a3a2-0f22a4b5bcbb)

Despite not importing any projected part correctly, I believe this is "closer to correct" because the pizza slices (and the studs!) now have "contiguous" textures. There are two problems:
- The pizza texture's UVs are *shifted* for some reason. I think I'm going to ignore this until it's the only problem.
![image](https://github.com/user-attachments/assets/d1fea233-80f4-4398-ab23-f9cd9e098553)
- More blatantly, all the projected textures are mirrored along Blender's Y-axis. The case with the rotated print is particularly interesting, because the texture being projected has the square off-center but orthogonal on a large transparent canvas, and yet the mirrored square still seems to be in the right position.
![image](https://github.com/user-attachments/assets/4c3ce6ae-e0b0-42ba-bab8-772fe026ada0)

**Comment by The0x539** [comment]
I think I've got it! The two problems were related somehow, cancelling out each other's translation issues. Mirroring the along the (blender Y, LDraw Z) axis was the trick, but the first thing I tried was literally doing `y = 1 - y` as in your advice, which didn't work. Applying the mirroring in the "transformed space" of `init_texture_transform`, however, works like a charm.
![image](https://github.com/user-attachments/assets/c27bc31b-0b6c-4d70-b864-b02e9c93f86b)

**Comment by ScanMountGoat** [comment]
That's awesome. I'd be happy to review a pull request with what you have done so far when you get a chance. It would be nice to figure out the bounding box and face thresholds, but that seems to be separate from the actual UV calculation.

**Comment by The0x539** [comment]
I expect those to be pretty easy, as well as switching the Python data to a packed format. If any of that stumps me, though, I'll call it a "day" and put the stuff that *does* work into a PR.

**Comment by ScanMountGoat** [comment]
![image](https://github.com/user-attachments/assets/76d5907a-6e90-4874-9c6a-e2a388051d1f)
pizza!

**Comment by tobblx** [comment]
Wow. I'm really proud of you guys. 
That was a hard work and needed deep investigations having so less public informations about it.
Looking forward to 0.43 :-)
Kudos!

**Comment by tobblx** [comment]
Hello, 
I just tested 0.4.3. Everything works great out of the box now. 
The direct import of the modified io-Files is also a big relief for me. 
Thank you very much.
